### PR TITLE
[CONTRIB] MPS DNN Dense

### PR DESCRIPTION
**Pull request by szha**

### CODE REVIEW

File: src/contrib/mps/gemm.mm [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      CHECK_EQ(C->ndim, 2);
+      CHECK(C->strides == nullptr);
+      CHECK(B->strides == nullptr);
+      CHECK(A->strides == nullptr);
+      CHECK(TypeMatch(A->dtype, kDLFloat, 32));
+      CHECK(TypeMatch(B->dtype, kDLFloat, 32));
+      CHECK(TypeMatch(C->dtype, kDLFloat, 32));
+      // Get Metal device API
+      auto func = runtime::Registry::Get("device_api.metal");
+      void *dev_handle = (*func)();
```
Comment (tqchen) [inline src/contrib/mps/gemm.mm]: device API getter need to put into a separate function, with the device handle as static variable.

File: src/contrib/mps/gemm.mm [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      runtime::metal::MetalWorkspace *metal_api =
+          static_cast<runtime::metal::MetalWorkspace *>(dev_handle);
+      // TODO(Check same device)
+      id<MTLDevice> dev = metal_api->GetDevice(A->ctx);
+      id<MTLCommandQueue> queue = metal_api->GetCommandQueue(A->ctx);
+      id<MTLCommandBuffer> cb = [queue commandBuffer];
+      // TODO(Check shape)
```
Comment (tqchen) [inline src/contrib/mps/gemm.mm]: add shape checks

File: src/contrib/mps/gemm.mm [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      CHECK(A->strides == nullptr);
+      CHECK(TypeMatch(A->dtype, kDLFloat, 32));
+      CHECK(TypeMatch(B->dtype, kDLFloat, 32));
+      CHECK(TypeMatch(C->dtype, kDLFloat, 32));
+      // Get Metal device API
+      auto func = runtime::Registry::Get("device_api.metal");
+      void *dev_handle = (*func)();
+      runtime::metal::MetalWorkspace *metal_api =
+          static_cast<runtime::metal::MetalWorkspace *>(dev_handle);
+      // TODO(Check same device)
```
Comment (tqchen) [inline src/contrib/mps/gemm.mm]: add device checks

File: src/contrib/mps/gemm.mm [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      id<MTLDevice> dev = metal_api->GetDevice(A->ctx);
+      id<MTLCommandQueue> queue = metal_api->GetCommandQueue(A->ctx);
+      id<MTLCommandBuffer> cb = [queue commandBuffer];
+      // TODO(Check shape)
+      NSUInteger M = A->shape[0];
+      NSUInteger N = B->shape[1];
+      NSUInteger K = B->shape[0];
+      // mps a
+      MPSMatrixDescriptor *descA = [MPSMatrixDescriptor
```
Comment (tqchen) [inline src/contrib/mps/gemm.mm]: do we need to cache descriptor?

File: src/contrib/mps/gemm.mm [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                                               interiorColumns:K
+                                                         alpha:1.0f
+                                                          beta:0.0f];
+      CHECK(sgemm != nil);
+      [sgemm encodeToCommandBuffer:cb
+                        leftMatrix:matrixA
+                       rightMatrix:matrixB
+                      resultMatrix:matrixC];
+      [cb commit];
+      [cb waitUntilCompleted];
```
Comment (tqchen) [inline src/contrib/mps/gemm.mm]: is there any need of wait until complete?

File: src/contrib/mps/gemm.mm (hunk 1 above)
Comment (szha) [reply src/contrib/mps/gemm.mm]: I'm following cudnn api and putting it in a thread-local entry.

File: src/contrib/mps/gemm.mm (hunk 3 above)
Comment (szha) [reply src/contrib/mps/gemm.mm]: added

File: src/contrib/mps/gemm.mm (hunk 2 above)
Comment (szha) [reply src/contrib/mps/gemm.mm]: added

File: src/contrib/mps/gemm.mm (hunk 4 above)
Comment (szha) [reply src/contrib/mps/gemm.mm]: I didn't find a setter for the descriptor in the metal api. I'm still learning the api so I may have missed a way to set it.

**Pull request merged**